modules/xfem/src/materials/LevelSetMultiMaterial.C

```diff
@@ -0,0 +1,159 @@
+  params.addRequiredParam<std::string>("prop_name", "Name for the computed material property.");
+  return params;
+}
+
+template <typename T, bool is_ad>
+LevelSetMultiMaterialTempl<T, is_ad>::LevelSetMultiMaterialTempl(const InputParameters & parameters)
+  : Material(parameters),
+    _level_set_var_names(getParam<std::vector<VariableName>>("level_set_vars")),
+    _nvars(_level_set_var_names.size()),
+    _level_set_var_numbers(_nvars),
+    _systems(_nvars),
+    _solutions(_nvars),
+    _keys(getParam<std::vector<std::string>>("base_name_keys")),
+    _vals(getParam<std::vector<std::string>>("base_name_vals")),
+
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _prop_name(getParam<std::string>("prop_name")),
+    _prop(declareGenericProperty<T, is_ad>(_base_name + _prop_name))
+{
+  FEProblemBase * fe_problem = dynamic_cast<FEProblemBase *>(&_subproblem);
+  if (!fe_problem)
+    mooseError("Problem casting _subproblem to FEProblemBase");
+  _xfem = MooseSharedNamespace::dynamic_pointer_cast<XFEM>(fe_problem->getXFEM());
+
+  for (unsigned int i = 0; i < _nvars; i++)
+  {
+    _level_set_var_numbers[i] = _subproblem
+                                    .getVariable(_tid,
+                                                 _level_set_var_names[i],
+                                                 Moose::VarKindType::VAR_ANY,
+                                                 Moose::VarFieldType::VAR_FIELD_STANDARD)
+                                    .number();
+    _systems[i] = &_subproblem.getSystem(_level_set_var_names[i]);
+    _solutions[i] = _systems[i]->current_local_solution.get();
+  }
+
+  for (unsigned int i = 0; i < _keys.size(); i++)
+  {
+    if (_keys[i].size() != _nvars)
+      mooseError("The length of each base_name_key must be equal to the number of level set "
+                 "variables which is ",
+                 _nvars,
+                 ", got key ",
+                 _keys[i]);
+    unsigned int key = encodeLevelSetKey(_keys[i]);
+    _base_name_map.emplace(key, _vals[i]);
+    _prop_map.emplace(key, &getGenericMaterialProperty<T, is_ad>(_vals[i] + "_" + _prop_name));
+  }
+}
+
+template <typename T, bool is_ad>
+void
+LevelSetMultiMaterialTempl<T, is_ad>::computeProperties()
+{
+  const Node * node = pickOnePhysicalNode();
+
+  // compute the key based on levelset variable signs
+  unsigned int key = 0;
+  for (unsigned int i = 0; i < _nvars; i++)
+  {
+    dof_id_type ls_dof_id = node->dof_number(_systems[i]->number(), _level_set_var_numbers[i], 0);
+    Number ls_node_value = (*_solutions[i])(ls_dof_id);
+    if (ls_node_value > 0.0)
+      key += std::pow(2, i);
+  }
+
+  // We may run into situations where the key doesn't exist in the base_name_map. This may happen
+  // when the problem is not well defined, or the level-set variables are very nonlinear. Cutting
+  // the timestep may help, so we throw an exception in this case.
```

Comment:
I am not fully understand why cut the tilmestep would help. And what do you mean by problem is not well defined, and the level set variables are very nonlinear?

Reply:
I added a sentence in the comment.